crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs

```diff
@@ -0,0 +1,97 @@
+use ruff_diagnostics::Diagnostic;
+use ruff_diagnostics::Violation;
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::visitor;
+use ruff_python_ast::visitor::Visitor;
+use ruff_python_ast::{self as ast, Expr, Stmt, StmtFunctionDef};
+use ruff_text_size::TextRange;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for `return x` statements in functions, that also contain yield
+/// statements.
+///
+/// ## Why is this bad?
+/// Using `return x` in a generator function used to be syntactically invalid
+/// in Python 2. In Python 3 `return x` can be used in a generator as a return
+/// value in conjunction with yield from. Users coming from Python 2 may expect
+/// the old behavior which might lead to bugs. Use native async def coroutines
+/// or mark intentional return x usage with # noqa on the same line.
+///
+/// ## Example
+/// ```python
+/// def broken():
+///     if True:
+///         return [1, 2, 3]
+///
+///     yield 3
+///     yield 2
+///     yield 1
+/// ```
+#[violation]
+pub struct ReturnXInGenerator;
+
+impl Violation for ReturnXInGenerator {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Using `yield` together with `return x`. Use native `async def` coroutines or put a `# noqa` comment on this line if this was intentional.")
+    }
+}
+
+#[derive(Default)]
+struct ReturnXInGeneratorVisitor {
+    return_: Option<TextRange>,
+    has_yield: bool,
+
+    in_expr_statement: bool,
+}
+
+impl Visitor<'_> for ReturnXInGeneratorVisitor {
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        match stmt {
+            Stmt::FunctionDef(_) => {
+                // do not recurse into nested functions, as they are evaluated
+                // individually
+            }
+            Stmt::Return(ast::StmtReturn { value, range }) => {
+                if Option::is_some(value) {
+                    self.return_ = Some(*range);
+                }
+            }
+            _ => {
+                self.in_expr_statement = stmt.is_expr_stmt();
```

Comment:
I think I would do this by checking `Stmt::Expr` here, and then just introspecting the value directly to see if the value is `Expr::Yield`.

Reply:
Good idea, I updated the implementation to this.